# Premiers modèles prédictifs du nombre de vélibs d'un point de vue spatial

In [1]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns

from sklearn.model_selection import train_test_split
from sklearn.linear_model import LinearRegression
from sklearn.metrics import mean_squared_error, mean_absolute_error
from sklearn.metrics import accuracy_score
from sklearn.metrics import f1_score, confusion_matrix, accuracy_score, recall_score, precision_score
from sklearn.preprocessing import PolynomialFeatures
from sklearn.metrics import mean_squared_error
from sklearn.tree import DecisionTreeRegressor
from sklearn.ensemble import RandomForestRegressor
from sklearn.datasets import make_classification
from sklearn import linear_model
from math import sqrt

!pip install prettytable
from prettytable import PrettyTable

In [2]:
velib = pd.read_csv(r"velib_clean.csv")
velib.head()

station_id               10/19/20;18:01:51               10/19/20;18:06:52  \
0   213688169   {'mechanical': 0, 'ebike': 2}   {'mechanical': 0, 'ebike': 2}   
1    99950133  {'mechanical': 41, 'ebike': 5}  {'mechanical': 41, 'ebike': 5}   
2   516709288   {'mechanical': 0, 'ebike': 1}   {'mechanical': 0, 'ebike': 1}   
3       36255   {'mechanical': 0, 'ebike': 0}   {'mechanical': 0, 'ebike': 0}   
4    37815204   {'mechanical': 4, 'ebike': 1}   {'mechanical': 3, 'ebike': 1}   

                10/19/20;18:11:52               10/19/20;18:16:52  \
0   {'mechanical': 0, 'ebike': 2}   {'mechanical': 0, 'ebike': 3}   
1  {'mechanical': 42, 'ebike': 5}  {'mechanical': 42, 'ebike': 7}   
2   {'mechanical': 0, 'ebike': 1}   {'mechanical': 1, 'ebike': 1}   
3   {'mechanical': 0, 'ebike': 0}   {'mechanical': 0, 'ebike': 0}   
4   {'mechanical': 1, 'ebike': 1}   {'mechanical': 1, 'ebike': 1}   

                10/19/20;18:21:52               10/19/20;18:26:53  \
0   {'mechanical': 0, 'ebike': 3}   {'mechanical': 1, 'ebike': 2}   
1  {'mechanical': 41, 'ebike': 5}  {'mechanical': 40, 'ebike': 4}   
2   {'mechanical': 1, 'ebike': 1}   {'mechanical': 1, 'ebike': 1}   
3   {'mechanical': 0, 'ebike': 2}   {'mechanical': 0, 'ebike': 4}   
4   {'mechanical': 2, 'ebike': 1}   {'mechanical': 3, 'ebike': 1}   

                10/19/20;18:31:53               10/19/20;18:36:53  \
0   {'mechanical': 0, 'ebike': 1}   {'mechanical': 0, 'ebike': 1}   
1  {'mechanical': 38, 'ebike': 5}  {'mechanical': 39, 'ebike': 4}   
2   {'mechanical': 2, 'ebike': 1}   {'mechanical': 0, 'ebike': 1}   
3   {'mechanical': 0, 'ebike': 4}   {'mechanical': 0, 'ebike': 4}   
4   {'mechanical': 3, 'ebike': 2}   {'mechanical': 2, 'ebike': 0}   

                10/19/20;18:41:54  ...              11/03/20;16:01:46  \
0   {'mechanical': 0, 'ebike': 1}  ...  {'mechanical': 4, 'ebike': 6}   
1  {'mechanical': 40, 'ebike': 5}  ...  {'mechanical': 2, 'ebike': 0}   
2   {'mechanical': 0, 'ebike': 1}  ...  {'mechanical': 1, 'ebike': 2}   
3   {'mechanical': 1, 'ebike': 3}  ...  {'mechanical': 0, 'ebike': 3}   
4   {'mechanical': 2, 'ebike': 0}  ...  {'mechanical': 6, 'ebike': 0}   

               11/03/20;16:06:47              11/03/20;16:11:47  \
0  {'mechanical': 4, 'ebike': 6}  {'mechanical': 3, 'ebike': 6}   
1  {'mechanical': 2, 'ebike': 0}  {'mechanical': 2, 'ebike': 0}   
2  {'mechanical': 1, 'ebike': 2}  {'mechanical': 1, 'ebike': 2}   
3  {'mechanical': 0, 'ebike': 3}  {'mechanical': 0, 'ebike': 3}   
4  {'mechanical': 7, 'ebike': 0}  {'mechanical': 7, 'ebike': 0}   

               11/03/20;16:16:47              11/03/20;16:21:48  \
0  {'mechanical': 4, 'ebike': 7}  {'mechanical': 5, 'ebike': 7}   
1  {'mechanical': 2, 'ebike': 1}  {'mechanical': 2, 'ebike': 1}   
2  {'mechanical': 1, 'ebike': 2}  {'mechanical': 1, 'ebike': 2}   
3  {'mechanical': 0, 'ebike': 3}  {'mechanical': 0, 'ebike': 2}   
4  {'mechanical': 7, 'ebike': 1}  {'mechanical': 7, 'ebike': 1}   

               11/03/20;16:26:48              11/03/20;16:31:48  \
0  {'mechanical': 5, 'ebike': 7}  {'mechanical': 5, 'ebike': 8}   
1  {'mechanical': 2, 'ebike': 0}  {'mechanical': 2, 'ebike': 0}   
2  {'mechanical': 1, 'ebike': 3}  {'mechanical': 1, 'ebike': 2}   
3  {'mechanical': 0, 'ebike': 2}  {'mechanical': 0, 'ebike': 3}   
4  {'mechanical': 8, 'ebike': 1}  {'mechanical': 9, 'ebike': 0}   

                11/03/20;16:36:49              11/03/20;16:41:49  \
0   {'mechanical': 5, 'ebike': 8}  {'mechanical': 5, 'ebike': 8}   
1   {'mechanical': 2, 'ebike': 0}  {'mechanical': 2, 'ebike': 0}   
2   {'mechanical': 1, 'ebike': 3}  {'mechanical': 1, 'ebike': 3}   
3   {'mechanical': 0, 'ebike': 3}  {'mechanical': 0, 'ebike': 3}   
4  {'mechanical': 10, 'ebike': 0}  {'mechanical': 9, 'ebike': 0}   

               11/03/20;16:46:49  
0  {'mechanical': 5, 'ebike': 7}  
1  {'mechanical': 2, 'ebike': 0}  
2  {'mechanical': 1, 'ebike': 3}  
3  {'mechanical': 0, 'ebike': 2}  
4  {'mechanical': 9, 'ebike': 0}  

[5 

In [3]:
velib2 = pd.read_csv(r"velib.csv")
velib2.head()

Unnamed: 0  station_id               10/19/20;18:01:51  \
0           0   213688169   {'mechanical': 0, 'ebike': 2}   
1           1    99950133  {'mechanical': 41, 'ebike': 5}   
2           2   516709288   {'mechanical': 0, 'ebike': 1}   
3           3       36255   {'mechanical': 0, 'ebike': 0}   
4           4    37815204   {'mechanical': 4, 'ebike': 1}   

                10/19/20;18:06:52               10/19/20;18:11:52  \
0   {'mechanical': 0, 'ebike': 2}   {'mechanical': 0, 'ebike': 2}   
1  {'mechanical': 41, 'ebike': 5}  {'mechanical': 42, 'ebike': 5}   
2   {'mechanical': 0, 'ebike': 1}   {'mechanical': 0, 'ebike': 1}   
3   {'mechanical': 0, 'ebike': 0}   {'mechanical': 0, 'ebike': 0}   
4   {'mechanical': 3, 'ebike': 1}   {'mechanical': 1, 'ebike': 1}   

                10/19/20;18:16:52               10/19/20;18:21:52  \
0   {'mechanical': 0, 'ebike': 3}   {'mechanical': 0, 'ebike': 3}   
1  {'mechanical': 42, 'ebike': 7}  {'mechanical': 41, 'ebike': 5}   
2   {'mechanical': 1, 'ebike': 1}   {'mechanical': 1, 'ebike': 1}   
3   {'mechanical': 0, 'ebike': 0}   {'mechanical': 0, 'ebike': 2}   
4   {'mechanical': 1, 'ebike': 1}   {'mechanical': 2, 'ebike': 1}   

                10/19/20;18:26:53               10/19/20;18:31:53  \
0   {'mechanical': 1, 'ebike': 2}   {'mechanical': 0, 'ebike': 1}   
1  {'mechanical': 40, 'ebike': 4}  {'mechanical': 38, 'ebike': 5}   
2   {'mechanical': 1, 'ebike': 1}   {'mechanical': 2, 'ebike': 1}   
3   {'mechanical': 0, 'ebike': 4}   {'mechanical': 0, 'ebike': 4}   
4   {'mechanical': 3, 'ebike': 1}   {'mechanical': 3, 'ebike': 2}   

                10/19/20;18:36:53  ...               02/15/21;21:13:34  \
0   {'mechanical': 0, 'ebike': 1}  ...   {'mechanical': 0, 'ebike': 4}   
1  {'mechanical': 39, 'ebike': 4}  ...  {'mechanical': 19, 'ebike': 5}   
2   {'mechanical': 0, 'ebike': 1}  ...   {'mechanical': 1, 'ebike': 2}   
3   {'mechanical': 0, 'ebike': 4}  ...   {'mechanical': 4, 'ebike': 4}   
4   {'mechanical': 2, 'ebike': 0}  ...  {'mechanical': 16, 'ebike': 2}   

                02/15/21;21:18:35               02/15/21;21:23:35  \
0   {'mechanical': 0, 'ebike': 4}   {'mechanical': 0, 'ebike': 4}   
1  {'mechanical': 19, 'ebike': 5}  {'mechanical': 19, 'ebike': 5}   
2   {'mechanical': 1, 'ebike': 2}   {'mechanical': 1, 'ebike': 2}   
3   {'mechanical': 4, 'ebike': 4}   {'mechanical': 3, 'ebike': 5}   
4  {'mechanical': 16, 'ebike': 2}  {'mechanical': 16, 'ebike': 2}   

                02/15/21;21:28:35               02/15/21;21:33:36  \
0   {'mechanical': 0, 'ebike': 4}   {'mechanical': 0, 'ebike': 4}   
1  {'mechanical': 19, 'ebike': 5}  {'mechanical': 19, 'ebike': 5}   
2   {'mechanical': 1, 'ebike': 2}   {'mechanical': 1, 'ebike': 2}   
3   {'mechanical': 3, 'ebike': 5}   {'mechanical': 3, 'ebike': 6}   
4  {'mechanical': 16, 'ebike': 3}  {'mechanical': 16, 'ebike': 3}   

                02/15/21;21:38:36               02/15/21;21:43:36  \
0   {'mechanical': 1, 'ebike': 4}   {'mechanical': 1, 'ebike': 4}   
1  {'mechanical': 19, 'ebike': 5}  {'mechanical': 19, 'ebike': 5}   
2   {'mechanical': 1, 'ebike': 2}   {'mechanical': 1, 'ebike': 2}   
3   {'mechanical': 3, 'ebike': 6}   {'mechanical': 3, 'ebike': 7}   
4  {'mechanical': 16, 'ebike': 3}  {'mechanical': 16, 'ebike': 3}   

                02/15/21;21:48:36               02/15/21;21:53:37  \
0   {'mechanical': 1, 'ebike': 4}   {'mechanical': 1, 'ebike': 4}   
1  {'mechanical': 19, 'ebike': 5}  {'mechanical': 19, 'ebike': 5}   
2   {'mechanical': 2, 'ebike': 3}   {'mechanical': 2, 'ebike': 2}   
3   {'mechanical': 2, 'ebike': 8}   {'mechanical': 2, 'ebike': 7}   
4  {'mechanical': 17, 'ebike': 3}  {'mechanical': 17, 'ebike': 4}   

                02/15/21;21:58:37  
0   {'mechanical': 1, 'ebike': 4}  
1  {'mechanical': 19, 'ebike': 5}  
2   {'mechanical': 2, 'ebike': 2}  
3   {'mechanical': 2, 'ebike': 7}  
4  {'mechanical': 16, 'ebike': 5}  

[5 rows x 33327 columns]

### Mise en forme des données

In [3]:
# Création d'une fonction pour convertir le format des données en un dictionnaire.
def to_dictionary(jour):
    """transformer le format en dico"""
    try :
        dictionary = eval(jour)
    except:
        dictionary = {'mechanical': 0, 'ebike': 0}
    
    return dictionary

# Mise en forme du dataframe velib en mettant le nombre de vélos dans un dictionnaire
col_vide='[{}, {}]'
colonnes_vides=[]
for j in velib.columns:
    if velib[j][0]==col_vide:
        print(j)
        colonnes_vides.append(j)
velib=velib.drop(colonnes_vides,axis=1)

# On applique la fonction pour convertir en dictionnaire
longueur_du_dataframe = len(velib.iloc[1])
for i,j in enumerate(velib.columns[1:]):
    velib[j] = [to_dictionary(i) for i in velib[j]]

In [4]:
velib.head()

station_id               10/19/20;18:01:51               10/19/20;18:06:52  \
0   213688169   {'mechanical': 0, 'ebike': 2}   {'mechanical': 0, 'ebike': 2}   
1    99950133  {'mechanical': 41, 'ebike': 5}  {'mechanical': 41, 'ebike': 5}   
2   516709288   {'mechanical': 0, 'ebike': 1}   {'mechanical': 0, 'ebike': 1}   
3       36255   {'mechanical': 0, 'ebike': 0}   {'mechanical': 0, 'ebike': 0}   
4    37815204   {'mechanical': 4, 'ebike': 1}   {'mechanical': 3, 'ebike': 1}   

                10/19/20;18:11:52               10/19/20;18:16:52  \
0   {'mechanical': 0, 'ebike': 2}   {'mechanical': 0, 'ebike': 3}   
1  {'mechanical': 42, 'ebike': 5}  {'mechanical': 42, 'ebike': 7}   
2   {'mechanical': 0, 'ebike': 1}   {'mechanical': 1, 'ebike': 1}   
3   {'mechanical': 0, 'ebike': 0}   {'mechanical': 0, 'ebike': 0}   
4   {'mechanical': 1, 'ebike': 1}   {'mechanical': 1, 'ebike': 1}   

                10/19/20;18:21:52               10/19/20;18:26:53  \
0   {'mechanical': 0, 'ebike': 3}   {'mechanical': 1, 'ebike': 2}   
1  {'mechanical': 41, 'ebike': 5}  {'mechanical': 40, 'ebike': 4}   
2   {'mechanical': 1, 'ebike': 1}   {'mechanical': 1, 'ebike': 1}   
3   {'mechanical': 0, 'ebike': 2}   {'mechanical': 0, 'ebike': 4}   
4   {'mechanical': 2, 'ebike': 1}   {'mechanical': 3, 'ebike': 1}   

                10/19/20;18:31:53               10/19/20;18:36:53  \
0   {'mechanical': 0, 'ebike': 1}   {'mechanical': 0, 'ebike': 1}   
1  {'mechanical': 38, 'ebike': 5}  {'mechanical': 39, 'ebike': 4}   
2   {'mechanical': 2, 'ebike': 1}   {'mechanical': 0, 'ebike': 1}   
3   {'mechanical': 0, 'ebike': 4}   {'mechanical': 0, 'ebike': 4}   
4   {'mechanical': 3, 'ebike': 2}   {'mechanical': 2, 'ebike': 0}   

                10/19/20;18:41:54  ...              11/03/20;16:01:46  \
0   {'mechanical': 0, 'ebike': 1}  ...  {'mechanical': 4, 'ebike': 6}   
1  {'mechanical': 40, 'ebike': 5}  ...  {'mechanical': 2, 'ebike': 0}   
2   {'mechanical': 0, 'ebike': 1}  ...  {'mechanical': 1, 'ebike': 2}   
3   {'mechanical': 1, 'ebike': 3}  ...  {'mechanical': 0, 'ebike': 3}   
4   {'mechanical': 2, 'ebike': 0}  ...  {'mechanical': 6, 'ebike': 0}   

               11/03/20;16:06:47              11/03/20;16:11:47  \
0  {'mechanical': 4, 'ebike': 6}  {'mechanical': 3, 'ebike': 6}   
1  {'mechanical': 2, 'ebike': 0}  {'mechanical': 2, 'ebike': 0}   
2  {'mechanical': 1, 'ebike': 2}  {'mechanical': 1, 'ebike': 2}   
3  {'mechanical': 0, 'ebike': 3}  {'mechanical': 0, 'ebike': 3}   
4  {'mechanical': 7, 'ebike': 0}  {'mechanical': 7, 'ebike': 0}   

               11/03/20;16:16:47              11/03/20;16:21:48  \
0  {'mechanical': 4, 'ebike': 7}  {'mechanical': 5, 'ebike': 7}   
1  {'mechanical': 2, 'ebike': 1}  {'mechanical': 2, 'ebike': 1}   
2  {'mechanical': 1, 'ebike': 2}  {'mechanical': 1, 'ebike': 2}   
3  {'mechanical': 0, 'ebike': 3}  {'mechanical': 0, 'ebike': 2}   
4  {'mechanical': 7, 'ebike': 1}  {'mechanical': 7, 'ebike': 1}   

               11/03/20;16:26:48              11/03/20;16:31:48  \
0  {'mechanical': 5, 'ebike': 7}  {'mechanical': 5, 'ebike': 8}   
1  {'mechanical': 2, 'ebike': 0}  {'mechanical': 2, 'ebike': 0}   
2  {'mechanical': 1, 'ebike': 3}  {'mechanical': 1, 'ebike': 2}   
3  {'mechanical': 0, 'ebike': 2}  {'mechanical': 0, 'ebike': 3}   
4  {'mechanical': 8, 'ebike': 1}  {'mechanical': 9, 'ebike': 0}   

                11/03/20;16:36:49              11/03/20;16:41:49  \
0   {'mechanical': 5, 'ebike': 8}  {'mechanical': 5, 'ebike': 8}   
1   {'mechanical': 2, 'ebike': 0}  {'mechanical': 2, 'ebike': 0}   
2   {'mechanical': 1, 'ebike': 3}  {'mechanical': 1, 'ebike': 3}   
3   {'mechanical': 0, 'ebike': 3}  {'mechanical': 0, 'ebike': 3}   
4  {'mechanical': 10, 'ebike': 0}  {'mechanical': 9, 'ebike': 0}   

               11/03/20;16:46:49  
0  {'mechanical': 5, 'ebike': 7}  
1  {'mechanical': 2, 'ebike': 0}  
2  {'mechanical': 1, 'ebike': 3}  
3  {'mechanical': 0, 'ebike': 2}  
4  {'mechanical': 9, 'ebike': 0}  

[5 

In [5]:
def to_str(var):
    if type(var) is list:
        return str(var)[1:-1] # list
    if type(var) is np.ndarray:
        try:
            return str(list(var[0]))[1:-1] # numpy 1D array
        except TypeError:
            return str(list(var))[1:-1] # numpy sequence
    return str(var) # everything else

In [6]:
toutes_les_stations = []
for i in range(len(velib)):
    toutes_les_stations.append(np.str(velib['station_id'][i]))

# Prédiction du nombre de vélos à la date t+1 en entrainant l'algorithme à la date t : exemple du 1er novembre à 10h pour prédire le 2 novembre à 10h. 

### Construction du dataframe pour le Machine Learning

In [5]:
df_velib = velib.set_index('station_id').transpose()
df_velib

station_id                             213688169  \
10/19/20;18:01:51  {'mechanical': 0, 'ebike': 2}   
10/19/20;18:06:52  {'mechanical': 0, 'ebike': 2}   
10/19/20;18:11:52  {'mechanical': 0, 'ebike': 2}   
10/19/20;18:16:52  {'mechanical': 0, 'ebike': 3}   
10/19/20;18:21:52  {'mechanical': 0, 'ebike': 3}   
...                                          ...   
11/03/20;16:26:48  {'mechanical': 5, 'ebike': 7}   
11/03/20;16:31:48  {'mechanical': 5, 'ebike': 8}   
11/03/20;16:36:49  {'mechanical': 5, 'ebike': 8}   
11/03/20;16:41:49  {'mechanical': 5, 'ebike': 8}   
11/03/20;16:46:49  {'mechanical': 5, 'ebike': 7}   

station_id                              99950133   \
10/19/20;18:01:51  {'mechanical': 41, 'ebike': 5}   
10/19/20;18:06:52  {'mechanical': 41, 'ebike': 5}   
10/19/20;18:11:52  {'mechanical': 42, 'ebike': 5}   
10/19/20;18:16:52  {'mechanical': 42, 'ebike': 7}   
10/19/20;18:21:52  {'mechanical': 41, 'ebike': 5}   
...                                           ...   
11/03/20;16:26:48   {'mechanical': 2, 'ebike': 0}   
11/03/20;16:31:48   {'mechanical': 2, 'ebike': 0}   
11/03/20;16:36:49   {'mechanical': 2, 'ebike': 0}   
11/03/20;16:41:49   {'mechanical': 2, 'ebike': 0}   
11/03/20;16:46:49   {'mechanical': 2, 'ebike': 0}   

station_id                             516709288  \
10/19/20;18:01:51  {'mechanical': 0, 'ebike': 1}   
10/19/20;18:06:52  {'mechanical': 0, 'ebike': 1}   
10/19/20;18:11:52  {'mechanical': 0, 'ebike': 1}   
10/19/20;18:16:52  {'mechanical': 1, 'ebike': 1}   
10/19/20;18:21:52  {'mechanical': 1, 'ebike': 1}   
...                                          ...   
11/03/20;16:26:48  {'mechanical': 1, 'ebike': 3}   
11/03/20;16:31:48  {'mechanical': 1, 'ebike': 2}   
11/03/20;16:36:49  {'mechanical': 1, 'ebike': 3}   
11/03/20;16:41:49  {'mechanical': 1, 'ebike': 3}   
11/03/20;16:46:49  {'mechanical': 1, 'ebike': 3}   

station_id                             36255      \
10/19/20;18:01:51  {'mechanical': 0, 'ebike': 0}   
10/19/20;18:06:52  {'mechanical': 0, 'ebike': 0}   
10/19/20;18:11:52  {'mechanical': 0, 'ebike': 0}   
10/19/20;18:16:52  {'mechanical': 0, 'ebike': 0}   
10/19/20;18:21:52  {'mechanical': 0, 'ebike': 2}   
...                                          ...   
11/03/20;16:26:48  {'mechanical': 0, 'ebike': 2}   
11/03/20;16:31:48  {'mechanical': 0, 'ebike': 3}   
11/03/20;16:36:49  {'mechanical': 0, 'ebike': 3}   
11/03/20;16:41:49  {'mechanical': 0, 'ebike': 3}   
11/03/20;16:46:49  {'mechanical': 0, 'ebike': 2}   

station_id                              37815204   \
10/19/20;18:01:51   {'mechanical': 4, 'ebike': 1}   
10/19/20;18:06:52   {'mechanical': 3, 'ebike': 1}   
10/19/20;18:11:52   {'mechanical': 1, 'ebike': 1}   
10/19/20;18:16:52   {'mechanical': 1, 'ebike': 1}   
10/19/20;18:21:52   {'mechanical': 2, 'ebike': 1}   
...                                           ...   
11/03/20;16:26:48   {'mechanical': 8, 'ebike': 1}   
11/03/20;16:31:48   {'mechanical': 9, 'ebike': 0}   
11/03/20;16:36:49  {'mechanical': 10, 'ebike': 0}   
11/03/20;16:41:49   {'mechanical': 9, 'ebike': 0}   
11/03/20;16:46:49   {'mechanical': 9, 'ebike': 0}   

station_id                               100769544  \
10/19/20;18:01:51  {'mechanical': 22, 'ebike': 20}   
10/19/20;18:06:52  {'mechanical': 22, 'ebike': 20}   
10/19/20;18:11:52  {'mechanical': 22, 'ebike': 20}   
10/19/20;18:16:52  {'mechanical': 22, 'ebike': 19}   
10/19/20;18:21:52  {'mechanical': 23, 'ebike': 19}   
...                                            ...   
11/03/20;16:26:48  {'mechanical': 22, 'ebike': 21}   
11/03/20;16:31:48  {'mechanical': 22, 'ebike': 20}   
11/03/20;16:36:49  {'mechanical': 22, 'ebike': 20}   
11/03/20;16:41:49  {'mechanical': 22, 'ebike': 19}   
11/03/20;16:46:49  {'mechanical': 22, 'ebike': 18}   

station_id                              85002689   \
10/19/20;18:01:51   {'mechanical': 2, 'ebike': 2}   
10/19/20;18:06:52   {'mechanical': 2, 'ebike': 2}   
10/19/20;18:11:52   {'mechanical': 2, 'ebik

On convertit les colonnes en string

In [6]:
def colonnes_string(df):
    """
    Convertir toutes les colonnes en string par simplicité
    """
    colonnes = []
    for i in df:
        if type(i) == str:
            colonnes.append(i)
        else:
            colonnes.append(str(i))

colonnes_string(df_velib)
df_velib

station_id                             213688169  \
10/19/20;18:01:51  {'mechanical': 0, 'ebike': 2}   
10/19/20;18:06:52  {'mechanical': 0, 'ebike': 2}   
10/19/20;18:11:52  {'mechanical': 0, 'ebike': 2}   
10/19/20;18:16:52  {'mechanical': 0, 'ebike': 3}   
10/19/20;18:21:52  {'mechanical': 0, 'ebike': 3}   
...                                          ...   
11/03/20;16:26:48  {'mechanical': 5, 'ebike': 7}   
11/03/20;16:31:48  {'mechanical': 5, 'ebike': 8}   
11/03/20;16:36:49  {'mechanical': 5, 'ebike': 8}   
11/03/20;16:41:49  {'mechanical': 5, 'ebike': 8}   
11/03/20;16:46:49  {'mechanical': 5, 'ebike': 7}   

station_id                              99950133   \
10/19/20;18:01:51  {'mechanical': 41, 'ebike': 5}   
10/19/20;18:06:52  {'mechanical': 41, 'ebike': 5}   
10/19/20;18:11:52  {'mechanical': 42, 'ebike': 5}   
10/19/20;18:16:52  {'mechanical': 42, 'ebike': 7}   
10/19/20;18:21:52  {'mechanical': 41, 'ebike': 5}   
...                                           ...   
11/03/20;16:26:48   {'mechanical': 2, 'ebike': 0}   
11/03/20;16:31:48   {'mechanical': 2, 'ebike': 0}   
11/03/20;16:36:49   {'mechanical': 2, 'ebike': 0}   
11/03/20;16:41:49   {'mechanical': 2, 'ebike': 0}   
11/03/20;16:46:49   {'mechanical': 2, 'ebike': 0}   

station_id                             516709288  \
10/19/20;18:01:51  {'mechanical': 0, 'ebike': 1}   
10/19/20;18:06:52  {'mechanical': 0, 'ebike': 1}   
10/19/20;18:11:52  {'mechanical': 0, 'ebike': 1}   
10/19/20;18:16:52  {'mechanical': 1, 'ebike': 1}   
10/19/20;18:21:52  {'mechanical': 1, 'ebike': 1}   
...                                          ...   
11/03/20;16:26:48  {'mechanical': 1, 'ebike': 3}   
11/03/20;16:31:48  {'mechanical': 1, 'ebike': 2}   
11/03/20;16:36:49  {'mechanical': 1, 'ebike': 3}   
11/03/20;16:41:49  {'mechanical': 1, 'ebike': 3}   
11/03/20;16:46:49  {'mechanical': 1, 'ebike': 3}   

station_id                             36255      \
10/19/20;18:01:51  {'mechanical': 0, 'ebike': 0}   
10/19/20;18:06:52  {'mechanical': 0, 'ebike': 0}   
10/19/20;18:11:52  {'mechanical': 0, 'ebike': 0}   
10/19/20;18:16:52  {'mechanical': 0, 'ebike': 0}   
10/19/20;18:21:52  {'mechanical': 0, 'ebike': 2}   
...                                          ...   
11/03/20;16:26:48  {'mechanical': 0, 'ebike': 2}   
11/03/20;16:31:48  {'mechanical': 0, 'ebike': 3}   
11/03/20;16:36:49  {'mechanical': 0, 'ebike': 3}   
11/03/20;16:41:49  {'mechanical': 0, 'ebike': 3}   
11/03/20;16:46:49  {'mechanical': 0, 'ebike': 2}   

station_id                              37815204   \
10/19/20;18:01:51   {'mechanical': 4, 'ebike': 1}   
10/19/20;18:06:52   {'mechanical': 3, 'ebike': 1}   
10/19/20;18:11:52   {'mechanical': 1, 'ebike': 1}   
10/19/20;18:16:52   {'mechanical': 1, 'ebike': 1}   
10/19/20;18:21:52   {'mechanical': 2, 'ebike': 1}   
...                                           ...   
11/03/20;16:26:48   {'mechanical': 8, 'ebike': 1}   
11/03/20;16:31:48   {'mechanical': 9, 'ebike': 0}   
11/03/20;16:36:49  {'mechanical': 10, 'ebike': 0}   
11/03/20;16:41:49   {'mechanical': 9, 'ebike': 0}   
11/03/20;16:46:49   {'mechanical': 9, 'ebike': 0}   

station_id                               100769544  \
10/19/20;18:01:51  {'mechanical': 22, 'ebike': 20}   
10/19/20;18:06:52  {'mechanical': 22, 'ebike': 20}   
10/19/20;18:11:52  {'mechanical': 22, 'ebike': 20}   
10/19/20;18:16:52  {'mechanical': 22, 'ebike': 19}   
10/19/20;18:21:52  {'mechanical': 23, 'ebike': 19}   
...                                            ...   
11/03/20;16:26:48  {'mechanical': 22, 'ebike': 21}   
11/03/20;16:31:48  {'mechanical': 22, 'ebike': 20}   
11/03/20;16:36:49  {'mechanical': 22, 'ebike': 20}   
11/03/20;16:41:49  {'mechanical': 22, 'ebike': 19}   
11/03/20;16:46:49  {'mechanical': 22, 'ebike': 18}   

station_id                              85002689   \
10/19/20;18:01:51   {'mechanical': 2, 'ebike': 2}   
10/19/20;18:06:52   {'mechanical': 2, 'ebike': 2}   
10/19/20;18:11:52   {'mechanical': 2, 'ebik

On ne regarde que le nombre total de vélos par station

In [7]:
def count_bikes(df):
    """
    Convertir les dictionnaires des différents Vélibs en un nombre de Vélibs disponibles à l'instant t.
    """
    for station in df.columns:
        for date in df.index:
            df[station][date] = df[station][date]['mechanical'] + df[station][date]['ebike']

count_bikes(df_velib)

In [9]:
df_velib

station_id        213688169 99950133  516709288 36255     37815204  100769544  \
10/19/20;18:01:51         2        46         1         0         5        42   
10/19/20;18:06:52         2        46         1         0         4        42   
10/19/20;18:11:52         2        47         1         0         2        42   
10/19/20;18:16:52         3        49         2         0         2        41   
10/19/20;18:21:52         3        46         2         2         3        42   
...                     ...       ...       ...       ...       ...       ...   
11/03/20;16:26:48        12         2         4         2         9        43   
11/03/20;16:31:48        13         2         3         3         9        42   
11/03/20;16:36:49        13         2         4         3        10        42   
11/03/20;16:41:49        13         2         4         3         9        41   
11/03/20;16:46:49        12         2         4         2         9        40   

station_id        85002689  54000559  85043758  123095125  ... 125881065  \
10/19/20;18:01:51         4         6         4         3  ...         3   
10/19/20;18:06:52         4         5         4         4  ...         4   
10/19/20;18:11:52         4         6         5         7  ...         2   
10/19/20;18:16:52         5         8         2         9  ...         2   
10/19/20;18:21:52         4         7         1        10  ...         3   
...                     ...       ...       ...       ...  ...       ...   
11/03/20;16:26:48        20        15        16        22  ...         1   
11/03/20;16:31:48        19        15        16        24  ...         1   
11/03/20;16:36:49        20        15        17        23  ...         3   
11/03/20;16:41:49        20        13        17        22  ...         2   
11/03/20;16:46:49        20        12        18        23  ...         2   

station_id        54000627  15377776  84961467  54000629  27415004  27415128   \
10/19/20;18:01:51         4         1        45        19         8         4   
10/19/20;18:06:52         4         1        45        18         8         3   
10/19/20;18:11:52         4         1        46        19         8         2   
10/19/20;18:16:52         5         0        45        19         9         1   
10/19/20;18:21:52         4         0        44        20         7         1   
...                     ...       ...       ...       ...       ...       ...   
11/03/20;16:26:48         7         4         3        20         9         5   
11/03/20;16:31:48         7         4         3        20         9         5   
11/03/20;16:36:49         8         5         7        21        10         5   
11/03/20;16:41:49         9         5         6        21        10         6   
11/03/20;16:46:49         9         5         7        21        10         5   

station_id        27414937  66507230  100227241  
10/19/20;18:01:51         3        19        24  
10/19/20;18:06:52         6        19        21  
10/19/20;18:11:52         7        19        22  
10/19/20;18:16:52         7        21        21  
10/19/20;18:21:52         9        21        17  
...                     ...       ...       ...  
11/03/20;16:26:48        23        14        27  
11/03/20;16:31:48        23        13        26  
11/03/20;16:36:49        23        13        27  
11/03/20;16:41:49        24        14        27  
11/03/20;16:46:49        23        16        28  

[4294 rows x 1398 columns]

**Préparation de la variable à prédire : le nombre de Vélibs dans la station Benjamin Godard - Victor Hugo (station n°213688169) 1 jour plus tard.**

Pour simplifier la tâche, on sait que il y a $12$x$24 = 288$ pointages par jour (toutes les 5 minutes)

In [43]:
df = df_velib.copy()
df['y'] = df[213688169]
df.reset_index(inplace=True)

# Formatage du dataframe pour que la colonne 'Y' contienne les données de la station cible 1 jour plus tard. 
for i in range(len(df)):
    try:
        df['y'][i] = df[213688169][i + 288]
    except KeyError:
        df.drop(i,0,inplace=True)

df=df.drop([213688169,'index'], axis = 1)
df

station_id 99950133 516709288 36255 37815204 100769544 85002689 54000559  \
0                46         1     0        5        42        4        6   
1                46         1     0        4        42        4        5   
2                47         1     0        2        42        4        6   
3                49         2     0        2        41        5        8   
4                46         2     2        3        42        4        7   
...             ...       ...   ...      ...       ...      ...      ...   
4001              9         5     4        7        28       21       18   
4002              8         5     2        7        28       20       18   
4003              8         5     2        7        31       19       19   
4004              8         5     2        8        31       19       19   
4005              8         5     1        8        33       19       19   

station_id 85043758 123095125 129026597  ... 54000627 15377776 84961467  \
0                 4         3        15  ...        4        1       45   
1                 4         4        14  ...        4        1       45   
2                 5         7        12  ...        4        1       46   
3                 2         9        12  ...        5        0       45   
4                 1        10         9  ...        4        0       44   
...             ...       ...       ...  ...      ...      ...      ...   
4001              7        11        19  ...       13        4        5   
4002              7        11        19  ...       14        3        5   
4003              7        12        19  ...       14        2        6   
4004              7        13        19  ...       15        2        6   
4005              7        13        19  ...       15        2        3   

station_id 54000629 27415004 27415128 27414937 66507230 100227241   y  
0                19        8        4        3       19        24   6  
1                18        8        3        6       19        21   6  
2                19        8        2        7       19        22   5  
3                19        9        1        7       21        21   5  
4                20        7        1        9       21        17   5  
...             ...      ...      ...      ...      ...       ...  ..  
4001             19        3        2       19       15        31  12  
4002             19        3        2       19       15        31  13  
4003             19        3        2       17       16        29  13  
4004             19        4        2       17       15        30  13  
4005             19        4        2       17       15        30  12  

[4006 rows x 1398 columns]

# Machine Learning

On entraine l'algorithme du 19 octobre jusqu'au 2 novembre. Et on prédit donc ce qui se passe entre le 20 octobre et le 3 novembre.

In [51]:
# Choix des features
stations = [i for i in df.columns if i != 'y']
X = df[stations]

# Préparation de la target
y = df['y']

from sklearn.model_selection import train_test_split

X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.3)

## Rappel MAE : 

MAE = mean($\lvert$ True - predicted $\rvert$)

In [58]:
# Régression linéaire simple
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.3)
regressor = LinearRegression()
regressor.fit(X_train, y_train)
y_prediction = regressor.predict(X_test)

print("Le MAE du modèle est "+str(round(mean_absolute_error(y_test, y_prediction),3))+" avec la régression linéaire")

Le MAE du modèle est 0.571 avec la régression linéaire


In [56]:
# Random Forest
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.3)
regressor = RandomForestRegressor()
regressor.fit(X_train, y_train)
y_prediction = regressor.predict(X_test)

print("Le MAE du modèle est "+str(round(mean_absolute_error(y_test, y_prediction),3))+" avec random forest")

Le MAE du modèle est 0.341 avec random forest


In [59]:
# Régression logistique ?
from sklearn.linear_model import LogisticRegression

y=y.astype('int')
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.3)

regressor = LogisticRegression(penalty='none',solver='newton-cg')
regressor.fit(X_train, y_train)
y_prediction = regressor.predict(X_test)

print("Le MAE du modèle est "+str(round(mean_absolute_error(y_test, y_prediction),3))+" avec régression logistique")

Le MAE du modèle est 0.317 avec régression logistique


## Sélection des stations les plus corrélées avec la target 

In [61]:
df = df.astype('float64')

In [68]:
heatmap = df.corr()
heatmap

station_id  99950133  516709288     36255  37815204  100769544  85002689  \
station_id                                                                 
99950133    1.000000  -0.195286  0.162100 -0.187678   0.611541 -0.177319   
516709288  -0.195286   1.000000  0.291150  0.396716  -0.207196  0.525433   
36255       0.162100   0.291150  1.000000  0.380271   0.150598  0.287062   
37815204   -0.187678   0.396716  0.380271  1.000000  -0.248167  0.507686   
100769544   0.611541  -0.207196  0.150598 -0.248167   1.000000 -0.260298   
...              ...        ...       ...       ...        ...       ...   
27415128   -0.056689   0.142942  0.114634  0.545498  -0.227218  0.293360   
27414937   -0.194966   0.520540  0.354176  0.717605  -0.337625  0.676606   
66507230   -0.104690  -0.034888 -0.110283 -0.105114  -0.078619 -0.143085   
100227241  -0.167219  -0.247694 -0.162988 -0.362688   0.183436 -0.380905   
y           0.068436   0.119396  0.008763  0.089050   0.223005  0.102308   

station_id  54000559  85043758  123095125  129026597  ...  54000627  15377776  \
station_id                                            ...                       
99950133   -0.334598 -0.223679  -0.204393   0.086788  ... -0.324523  0.025232   
516709288   0.279739  0.119680   0.242414  -0.200245  ...  0.513984 -0.140892   
36255       0.139330  0.119667   0.216087  -0.286006  ...  0.210502 -0.261493   
37815204    0.100615  0.502248   0.444494  -0.668032  ...  0.493721 -0.462698   
100769544  -0.302487 -0.417616  -0.353847   0.220972  ... -0.193849  0.201634   
...              ...       ...        ...        ...  ...       ...       ...   
27415128    0.209992  0.589388   0.257704  -0.583440  ...  0.353565 -0.206648   
27414937    0.310551  0.365913   0.472470  -0.444952  ...  0.549094 -0.428744   
66507230   -0.026049 -0.146478  -0.149499   0.263272  ... -0.013344  0.108291   
100227241  -0.178667 -0.135965  -0.393200   0.167123  ... -0.248515  0.239717   
y          -0.032470  0.016063  -0.165194  -0.101876  ...  0.235661  0.211848   

station_id  84961467  54000629  27415004  27415128  27414937  66507230  \
station_id                                                               
99950133    0.338678 -0.044211 -0.123873 -0.056689 -0.194966 -0.104690   
516709288  -0.468952  0.248628  0.215983  0.142942  0.520540 -0.034888   
36255      -0.253723  0.046560  0.417692  0.114634  0.354176 -0.110283   
37815204   -0.400822  0.226539  0.848933  0.545498  0.717605 -0.105114   
100769544   0.366349 -0.259098 -0.173361 -0.227218 -0.337625 -0.078619   
...              ...       ...       ...       ...       ...       ...   
27415128   -0.184490  0.214370  0.462206  1.000000  0.472193 -0.190974   
27414937   -0.551809  0.546129  0.642461  0.472193  1.000000  0.020199   
66507230    0.095784 -0.056796 -0.166057 -0.190974  0.020199  1.000000   
100227241   0.281107 -0.415353 -0.308554 -0.294577 -0.581496  0.148406   
y           0.065580 -0.001845 -0.013962  0.243408  0.024041 -0.168761   

station_id  100227241         y  
station_id                       
99950133    -0.167219  0.068436  
516709288   -0.247694  0.119396  
36255       -0.162988  0.008763  
37815204    -0.362688  0.089050  
100769544    0.183436  0.223005  
...               ...       ...  
27415128    -0.294577  0.243408  
27414937    -0.581496  0.024041  
66507230     0.148406 -0.168761  
100227241    1.000000  0.134075  
y            0.134075  1.000000  

[1398 rows x 1398 columns]

In [72]:
heatmap['y'][516709288]

0.11939606840184183

In [75]:
def choix_stations(df, target, seuil):
    """
    df : dataframe pour le ML
    target : variable à prédire (output)
    seuil : seuil minimum à partir duquel on accepte une variable dans la modélisation (en valeur absolue)
    Fonction qui renvoie la liste des features sélectionnées pour la modélisation
    """
    heatmap = df.corr()
    stations_correlees = []
    stations_possibles = [i for i in df.columns if i != target]

    for i in stations_possibles:
        if abs(heatmap[target][i]) > seuil:
            stations_correlees.append(i)
    return stations_correlees

In [67]:
df

station_id  99950133  516709288  36255  37815204  100769544  85002689  \
0               46.0        1.0    0.0       5.0       42.0       4.0   
1               46.0        1.0    0.0       4.0       42.0       4.0   
2               47.0        1.0    0.0       2.0       42.0       4.0   
3               49.0        2.0    0.0       2.0       41.0       5.0   
4               46.0        2.0    2.0       3.0       42.0       4.0   
...              ...        ...    ...       ...        ...       ...   
4001             9.0        5.0    4.0       7.0       28.0      21.0   
4002             8.0        5.0    2.0       7.0       28.0      20.0   
4003             8.0        5.0    2.0       7.0       31.0      19.0   
4004             8.0        5.0    2.0       8.0       31.0      19.0   
4005             8.0        5.0    1.0       8.0       33.0      19.0   

station_id  54000559  85043758  123095125  129026597  ...  54000627  15377776  \
0                6.0       4.0        3.0       15.0  ...       4.0       1.0   
1                5.0       4.0        4.0       14.0  ...       4.0       1.0   
2                6.0       5.0        7.0       12.0  ...       4.0       1.0   
3                8.0       2.0        9.0       12.0  ...       5.0       0.0   
4                7.0       1.0       10.0        9.0  ...       4.0       0.0   
...              ...       ...        ...        ...  ...       ...       ...   
4001            18.0       7.0       11.0       19.0  ...      13.0       4.0   
4002            18.0       7.0       11.0       19.0  ...      14.0       3.0   
4003            19.0       7.0       12.0       19.0  ...      14.0       2.0   
4004            19.0       7.0       13.0       19.0  ...      15.0       2.0   
4005            19.0       7.0       13.0       19.0  ...      15.0       2.0   

station_id  84961467  54000629  27415004  27415128  27414937  66507230  \
0               45.0      19.0       8.0       4.0       3.0      19.0   
1               45.0      18.0       8.0       3.0       6.0      19.0   
2               46.0      19.0       8.0       2.0       7.0      19.0   
3               45.0      19.0       9.0       1.0       7.0      21.0   
4               44.0      20.0       7.0       1.0       9.0      21.0   
...              ...       ...       ...       ...       ...       ...   
4001             5.0      19.0       3.0       2.0      19.0      15.0   
4002             5.0      19.0       3.0       2.0      19.0      15.0   
4003             6.0      19.0       3.0       2.0      17.0      16.0   
4004             6.0      19.0       4.0       2.0      17.0      15.0   
4005             3.0      19.0       4.0       2.0      17.0      15.0   

station_id  100227241     y  
0                24.0   6.0  
1                21.0   6.0  
2                22.0   5.0  
3                21.0   5.0  
4                17.0   5.0  
...               ...   ...  
4001             31.0  12.0  
4002             31.0  13.0  
4003             29.0  13.0  
4004             30.0  13.0  
4005             30.0  12.0  

[4006 rows x 1398 columns]

In [79]:
# On l'applique à notre jeu de données
stations_correlees = choix_stations(df,'y',0.5)
stations_correlees

[213688824, 977206088, 54000545, 213682190]

In [80]:
len(stations_correlees)

4

### Machine Learning avec des stations corrélées

In [86]:
X = df.loc[:,stations_correlees]
Y = df.loc[:,'y']

In [87]:
#Régression linéaire 

X_train, X_test, y_train, y_test = train_test_split(X, Y, test_size=0.3)

regressor = LinearRegression()
regressor.fit(X_train, y_train)
y_prediction = regressor.predict(X_test)

print("Le MAE du modèle est "+str(round(mean_absolute_error(y_test, y_prediction),3))+" avec la régression linéaire")

Le MAE du modèle est 1.696 avec la régression linéaire


In [88]:
# Random Forest

X_train, X_test, y_train, y_test = train_test_split(X, Y, test_size=0.3)
regressor = RandomForestRegressor()
regressor.fit(X_train, y_train)
y_prediction = regressor.predict(X_test)

print("Le MAE du modèle est "+str(round(mean_absolute_error(y_test, y_prediction),3))+" avec random forest")

Le MAE du modèle est 0.579 avec random forest


In [89]:
# Régression logistique ?

Y=Y.astype('int')
X_train, X_test, y_train, y_test = train_test_split(X, Y, test_size=0.3)

regressor = LogisticRegression(penalty='none',solver='newton-cg')
regressor.fit(X_train, y_train)
y_prediction = regressor.predict(X_test)

print("Le MAE du modèle est "+str(round(mean_absolute_error(y_test, y_prediction),3))+" avec régression logistique")

Le MAE du modèle est 1.421 avec régression logistique


# Test râté

### On entraine sur toutes nos données depuis le 20 octobre 2020 jusqu'au 1er novembre 2020 pour prédire le nombre de vélos dans une station aléatoire à chaque instant

In [39]:
velib = pd.read_csv("velib_clean.csv")
velib.head()

station_id               10/19/20;18:01:51               10/19/20;18:06:52  \
0   213688169   {'mechanical': 0, 'ebike': 2}   {'mechanical': 0, 'ebike': 2}   
1    99950133  {'mechanical': 41, 'ebike': 5}  {'mechanical': 41, 'ebike': 5}   
2   516709288   {'mechanical': 0, 'ebike': 1}   {'mechanical': 0, 'ebike': 1}   
3       36255   {'mechanical': 0, 'ebike': 0}   {'mechanical': 0, 'ebike': 0}   
4    37815204   {'mechanical': 4, 'ebike': 1}   {'mechanical': 3, 'ebike': 1}   

                10/19/20;18:11:52               10/19/20;18:16:52  \
0   {'mechanical': 0, 'ebike': 2}   {'mechanical': 0, 'ebike': 3}   
1  {'mechanical': 42, 'ebike': 5}  {'mechanical': 42, 'ebike': 7}   
2   {'mechanical': 0, 'ebike': 1}   {'mechanical': 1, 'ebike': 1}   
3   {'mechanical': 0, 'ebike': 0}   {'mechanical': 0, 'ebike': 0}   
4   {'mechanical': 1, 'ebike': 1}   {'mechanical': 1, 'ebike': 1}   

                10/19/20;18:21:52               10/19/20;18:26:53  \
0   {'mechanical': 0, 'ebike': 3}   {'mechanical': 1, 'ebike': 2}   
1  {'mechanical': 41, 'ebike': 5}  {'mechanical': 40, 'ebike': 4}   
2   {'mechanical': 1, 'ebike': 1}   {'mechanical': 1, 'ebike': 1}   
3   {'mechanical': 0, 'ebike': 2}   {'mechanical': 0, 'ebike': 4}   
4   {'mechanical': 2, 'ebike': 1}   {'mechanical': 3, 'ebike': 1}   

                10/19/20;18:31:53               10/19/20;18:36:53  \
0   {'mechanical': 0, 'ebike': 1}   {'mechanical': 0, 'ebike': 1}   
1  {'mechanical': 38, 'ebike': 5}  {'mechanical': 39, 'ebike': 4}   
2   {'mechanical': 2, 'ebike': 1}   {'mechanical': 0, 'ebike': 1}   
3   {'mechanical': 0, 'ebike': 4}   {'mechanical': 0, 'ebike': 4}   
4   {'mechanical': 3, 'ebike': 2}   {'mechanical': 2, 'ebike': 0}   

                10/19/20;18:41:54  ...              11/03/20;16:01:46  \
0   {'mechanical': 0, 'ebike': 1}  ...  {'mechanical': 4, 'ebike': 6}   
1  {'mechanical': 40, 'ebike': 5}  ...  {'mechanical': 2, 'ebike': 0}   
2   {'mechanical': 0, 'ebike': 1}  ...  {'mechanical': 1, 'ebike': 2}   
3   {'mechanical': 1, 'ebike': 3}  ...  {'mechanical': 0, 'ebike': 3}   
4   {'mechanical': 2, 'ebike': 0}  ...  {'mechanical': 6, 'ebike': 0}   

               11/03/20;16:06:47              11/03/20;16:11:47  \
0  {'mechanical': 4, 'ebike': 6}  {'mechanical': 3, 'ebike': 6}   
1  {'mechanical': 2, 'ebike': 0}  {'mechanical': 2, 'ebike': 0}   
2  {'mechanical': 1, 'ebike': 2}  {'mechanical': 1, 'ebike': 2}   
3  {'mechanical': 0, 'ebike': 3}  {'mechanical': 0, 'ebike': 3}   
4  {'mechanical': 7, 'ebike': 0}  {'mechanical': 7, 'ebike': 0}   

               11/03/20;16:16:47              11/03/20;16:21:48  \
0  {'mechanical': 4, 'ebike': 7}  {'mechanical': 5, 'ebike': 7}   
1  {'mechanical': 2, 'ebike': 1}  {'mechanical': 2, 'ebike': 1}   
2  {'mechanical': 1, 'ebike': 2}  {'mechanical': 1, 'ebike': 2}   
3  {'mechanical': 0, 'ebike': 3}  {'mechanical': 0, 'ebike': 2}   
4  {'mechanical': 7, 'ebike': 1}  {'mechanical': 7, 'ebike': 1}   

               11/03/20;16:26:48              11/03/20;16:31:48  \
0  {'mechanical': 5, 'ebike': 7}  {'mechanical': 5, 'ebike': 8}   
1  {'mechanical': 2, 'ebike': 0}  {'mechanical': 2, 'ebike': 0}   
2  {'mechanical': 1, 'ebike': 3}  {'mechanical': 1, 'ebike': 2}   
3  {'mechanical': 0, 'ebike': 2}  {'mechanical': 0, 'ebike': 3}   
4  {'mechanical': 8, 'ebike': 1}  {'mechanical': 9, 'ebike': 0}   

                11/03/20;16:36:49              11/03/20;16:41:49  \
0   {'mechanical': 5, 'ebike': 8}  {'mechanical': 5, 'ebike': 8}   
1   {'mechanical': 2, 'ebike': 0}  {'mechanical': 2, 'ebike': 0}   
2   {'mechanical': 1, 'ebike': 3}  {'mechanical': 1, 'ebike': 3}   
3   {'mechanical': 0, 'ebike': 3}  {'mechanical': 0, 'ebike': 3}   
4  {'mechanical': 10, 'ebike': 0}  {'mechanical': 9, 'ebike': 0}   

               11/03/20;16:46:49  
0  {'mechanical': 5, 'ebike': 7}  
1  {'mechanical': 2, 'ebike': 0}  
2  {'mechanical': 1, 'ebike': 3}  
3  {'mechanical': 0, 'ebike': 2}  
4  {'mechanical': 9, 'ebike': 0}  

[5 

### Préprocessing des données

In [40]:
# Création d'une fonction pour convertir le format des données en un dictionnaire.
def to_dictionary(jour):
    """transformer le format en dico"""
    try :
        dictionary = eval(jour)
    except:
        dictionary = {'mechanical': 0, 'ebike': 0}
    
    return dictionary

# Mise en forme du dataframe velib en mettant le nombre de vélos dans un dictionnaire
col_vide='[{}, {}]'
colonnes_vides=[]
for j in velib.columns:
    if velib[j][0]==col_vide:
        print(j)
        colonnes_vides.append(j)
velib=velib.drop(colonnes_vides,axis=1)

# On applique la fonction pour convertir en dictionnaire
longueur_du_dataframe = len(velib.iloc[1])
for i,j in enumerate(velib.columns[1:]):
    velib[j] = [to_dictionary(i) for i in velib[j]]

# On copie le dataframe pour l'utiliser ensuite
tous_les_velibs = velib.copy()

On isole les colonnes qui nous intéressent

In [101]:
colonnes_toutes_les_dates = [] # Colonnes correspondant à chaque instant jusqu'à la veille de la journée cible
colonnes_derniere_date = ['station_id'] # Colonnes correspondant à la journée cible
date1 = "10/19/20"
date_target = "11/02/20"
date3 = "11/03/20"

for i in tous_les_velibs:
    if (date1 not in i) and (date_target not in i) and (date3 not in i):
        colonnes_toutes_les_dates.append(i)
    if (date_target in i):
        colonnes_derniere_date.append(i)

In [108]:
# Vérification
len(colonnes_toutes_les_dates)

3736

On a cette fois 3736 lignes qui correspondent à autant d'intervalles de 5 minutes

In [107]:
# Vérification
len(colonnes_derniere_date)

287

On utilise des dataframes pour les dates qui nous intéressent

In [126]:
df_t1 = velib[colonnes_toutes_les_dates].set_index('station_id').transpose()
df_t2 = velib[colonnes_derniere_date].set_index('station_id').transpose()

On s'intéresse ensuite au nombre total de vélos par station

In [127]:
def count_bikes(df):
    for station in df:
        for i in range(len(df)):
            # On parcourt les lignes et on ne garde que le nombre total de vélibs à l'heure souhaitée
            df[station][i] = df[station][i]['mechanical'] + df[station][i]['ebike']

In [128]:
count_bikes(df_t1)
count_bikes(df_t2)

In [132]:
df_t1

station_id  level_0 213688169 99950133 516709288 36255 37815204 100769544  \
0                 0         3       52         9    10       24        23   
1                 1         3       52         9    10       24        23   
2                 2         3       52         9    10       24        23   
3                 3         3       52         9    10       24        23   
4                 4         3       52         9    10       24        23   
...             ...       ...      ...       ...   ...      ...       ...   
3730           3730         8       30        12     2       16        28   
3731           3731         8       30        12     2       16        28   
3732           3732         8       30        12     2       16        28   
3733           3733         8       30        12     2       16        28   
3734           3734         8       30        12     2       16        28   

station_id 85002689 54000559 85043758  ... 125881065 54000627 15377776  \
0                31        7       14  ...         8       13        0   
1                31        7       14  ...         8       13        0   
2                31        7       14  ...         8       13        0   
3                31        7       14  ...         8       13        0   
4                31        7       14  ...         8       13        0   
...             ...      ...      ...  ...       ...      ...      ...   
3730             28       13        8  ...         5       19        4   
3731             28       13        8  ...         5       19        4   
3732             28       13        8  ...         5       19        4   
3733             28       13        8  ...         5       19        4   
3734             28       13        8  ...         5       19        4   

station_id 84961467 54000629 27415004 27415128 27414937 66507230 100227241  
0                18       19       33       16       35       12         7  
1                18       19       33       16       35       12         7  
2                18       19       33       16       35       12         7  
3                18       19       33       16       35       12         7  
4                18       19       33       16       35       12         7  
...             ...      ...      ...      ...      ...      ...       ...  
3730             24       15       13       12       19       12        15  
3731             24       15       13       12       19       12        14  
3732             24       15       13       12       19       12        14  
3733             24       15       13       12       18       12        14  
3734             24       15       13       12       18       12        14  

[3735 rows x 1399 columns]

In [133]:
toutes_les_stations = []
for i in range(len(velib)):
    toutes_les_stations.append(np.str(velib['station_id'][i]))

# On reformate les données pour les manipuler
df_t1.reset_index(inplace=True)
df_t2.reset_index(inplace=True)

# On supprime les colonnes inutiles
df_t1 = df_t1.drop(['index'], axis = 1)
df_t2 = df_t2.drop(['index'], axis = 1)

In [136]:
# Suppression des autres colonnes inutiles
df_t1 = df_t1.drop(['level_0'], axis = 1)
df_t2 = df_t2.drop(['level_0'], axis = 1)

# On convertit le nom des colonnes en string
df_t1.columns = toutes_les_stations
df_t2.columns = toutes_les_stations

### Construction variable à prédire Y : une station random à la date t+1 i.e. le 2 novembre

On choisit aléatoirement la station numéro **213688169**, ce qui correspond à la station **Benjamin Godard - Victor Hugo** et qui dispose d'une capacité maximale de **35 vélibs**

In [137]:
y = df_t2[['213688169']]
y.columns = ['y']
y.head()

y
0  8
1  8
2  8
3  8
4  8

### Machine Learning sans tuning

In [144]:
df_t1

213688169 99950133 516709288 36255 37815204 100769544 85002689 54000559  \
0            3       52         9    10       24        23       31        7   
1            3       52         9    10       24        23       31        7   
2            3       52         9    10       24        23       31        7   
3            3       52         9    10       24        23       31        7   
4            3       52         9    10       24        23       31        7   
...        ...      ...       ...   ...      ...       ...      ...      ...   
3730         8       30        12     2       16        28       28       13   
3731         8       30        12     2       16        28       28       13   
3732         8       30        12     2       16        28       28       13   
3733         8       30        12     2       16        28       28       13   
3734         8       30        12     2       16        28       28       13   

     85043758 123095125  ... 125881065 54000627 15377776 84961467 54000629  \
0          14        36  ...         8       13        0       18       19   
1          14        36  ...         8       13        0       18       19   
2          14        36  ...         8       13        0       18       19   
3          14        36  ...         8       13        0       18       19   
4          14        36  ...         8       13        0       18       19   
...       ...       ...  ...       ...      ...      ...      ...      ...   
3730        8        30  ...         5       19        4       24       15   
3731        8        30  ...         5       19        4       24       15   
3732        8        30  ...         5       19        4       24       15   
3733        8        30  ...         5       19        4       24       15   
3734        8        31  ...         5       19        4       24       15   

     27415004 27415128 27414937 66507230 100227241  
0          33       16       35       12         7  
1          33       16       35       12         7  
2          33       16       35       12         7  
3          33       16       35       12         7  
4          33       16       35       12         7  
...       ...      ...      ...      ...       ...  
3730       13       12       19       12        15  
3731       13       12       19       12        14  
3732       13       12       19       12        14  
3733       13       12       18       12        14  
3734       13       12       18       12        14  

[3735 rows x 1398 columns]

In [145]:
df_t2

213688169 99950133 516709288 36255 37815204 100769544 85002689 54000559  \
0           8       30        12     2       16        28       28       13   
1           8       30        11     2       16        28       28       13   
2           8       30        11     2       16        28       28       13   
3           8       30        11     2       16        28       28       13   
4           8       30        11     2       16        28       28       13   
..        ...      ...       ...   ...      ...       ...      ...      ...   
281         9        8         6     3       13        23       27       15   
282         8        8         6     3       13        23       27       15   
283         9        8         6     3       13        23       27       15   
284         9        8         6     3       13        23       27       15   
285         9        8         6     3       13        23       27       15   

    85043758 123095125  ... 125881065 54000627 15377776 84961467 54000629  \
0          8        32  ...         5       19        4       24       15   
1          8        32  ...         5       19        4       24       15   
2          8        32  ...         5       19        4       24       15   
3          8        32  ...         5       19        4       24       15   
4          8        32  ...         5       19        4       24       15   
..       ...       ...  ...       ...      ...      ...      ...      ...   
281       16        24  ...        18       20        0       12       23   
282       16        22  ...        18       20        0       13       23   
283       16        22  ...        18       20        0       13       23   
284       16        22  ...        18       20        0       13       23   
285       16        22  ...        18       20        0       13       23   

    27415004 27415128 27414937 66507230 100227241  
0         13       12       18       12        14  
1         13       12       18       12        14  
2         13       12       17       12        14  
3         13       12       17       12        14  
4         13       12       17       12        14  
..       ...      ...      ...      ...       ...  
281       24        2       31        7        20  
282       24        2       31        7        20  
283       24        2       31        6        20  
284       24        2       31        6        20  
285       24        2       31        6        20  

[286 rows x 1398 columns]

In [146]:
# Choix des features et de la target
stations_features = toutes_les_stations
stations_features = [i for i in stations_features if i != '213688169']
X = df_t1[stations_features]

y = df_t1[['213688169']]

from sklearn.model_selection import train_test_split

X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.3)

### Méthode "naïve"

- Regarder si ce qui se passe à une date t est la même chose au même instant à la date t-1

Pour l'instant on recommence la procédure :

In [173]:
colonnes_t1 = ['station_id'] # Colonnes correspondant à chaque instant jusqu'à la veille de la journée cible
colonnes_t2 = ['station_id'] # Colonnes correspondant à la journée cible
datet1 = "10/31/20"
datet2 = "11/01/20"

for i in tous_les_velibs.columns:
    if (datet1 in i) :
        colonnes_t1.append(i)
    if (datet2 in i):
        colonnes_t2.append(i)

In [178]:
dft1 = velib[colonnes_t1].set_index('station_id').transpose()
dft2 = velib[colonnes_t2].set_index('station_id').transpose()

for station in dft1:
        for i in range(len(dft1)):
            # On parcourt les lignes et on ne garde que le nombre total de vélibs à l'heure souhaitée
            dft1[station][i] = dft1[station][i]['mechanical'] + dft1[station][i]['ebike']
            dft2[station][i] = dft2[station][i]['mechanical'] + dft2[station][i]['ebike']

In [223]:
#dft1.reset_index(inplace = True)
#dft2.reset_index(inplace = True)

#df1 = dft1.drop(['level_0'], axis = 1)
#df2 = dft2.drop(['level_0'], axis = 1)

#df_naif = df2[213688169]-df1[213688169]

dft1 = dft1.drop(['level_0','index'], axis = 1)
dft2 = dft2.drop(['level_0','index'], axis = 1)

In [225]:
t1 = dft1[[213688169]]
t2 = dft2[[213688169]]

In [238]:
df_naif = t2-t1
df_naif.head()

station_id 213688169
0                  0
1                 -1
2                 -1
3                 -2
4                 -2

In [240]:
from statistics import mean

In [243]:
#Calcul du MAE
df_naif = abs(df_naif)
df_naif.mean()

station_id
213688169    1.573427
dtype: float64

# Rappel MAE : 

MAE = mean($\lvert$ True - predicted $\rvert$)

- Rappel : on a supposé que le nombre de vélibs disponibles était exactement le même à la même heure que le jour précédent.

On obtient un MAE de 1.57 avec une méthode "naïve", donc une moyenne d'erreur de 1.57. **Cela semble peu (et donc plutôt bien!) dans la mesure où la station peut accueillir jusqu'à 35 vélibs. **

### Régression linéaire

In [149]:
# Régression linéaire simple
regressor = LinearRegression()
regressor.fit(X_train, y_train)
y_prediction = regressor.predict(X_test)

print("Le MAE du modèle est "+str(round(mean_absolute_error(y_test, y_prediction),3))+" avec la régression linéaire")

Le MAE du modèle est 0.584 avec la régression linéaire


On obtient de meilleurs résultats avec une régression linéaire simple sur cette station

### Random Forest

In [155]:
# Random Forest
regressor = RandomForestRegressor()
regressor.fit(X_train, y_train)
y_prediction = regressor.predict(X_test)

print("Le MAE du modèle est "+str(round(mean_absolute_error(y_test, y_prediction),3))+" avec random forest")

Le MAE du modèle est 0.324 avec random forest


In [151]:
y_prediction

array([4.13, 7.34, 2.56, ..., 0.55, 2.19, 3.05])

### Machine Learning avec tuning

In [252]:
df_ml = df_t1
df_ml = df_ml.astype('float64')

In [253]:
df_ml.corr()

213688169  99950133  516709288     36255  37815204  100769544  \
213688169   1.000000  0.025269   0.128240  0.082420  0.077572   0.187565   
99950133    0.025269  1.000000  -0.211264  0.094781 -0.292633   0.636175   
516709288   0.128240 -0.211264   1.000000  0.323731  0.412742  -0.178926   
36255       0.082420  0.094781   0.323731  1.000000  0.343862   0.139795   
37815204    0.077572 -0.292633   0.412742  0.343862  1.000000  -0.287526   
...              ...       ...        ...       ...       ...        ...   
27415004    0.121414 -0.219705   0.225187  0.381771  0.842328  -0.204427   
27415128   -0.088361 -0.062054   0.100604  0.131438  0.580039  -0.205263   
27414937    0.059033 -0.244888   0.526938  0.341646  0.739894  -0.342377   
66507230    0.104548 -0.090570  -0.021895 -0.093292 -0.059184  -0.075173   
100227241  -0.038691 -0.099520  -0.227173 -0.124456 -0.320766   0.189247   

           85002689  54000559  85043758  123095125  ...  125881065  54000627  \
213688169  0.231712  0.008926  0.037140  -0.030639  ...  -0.041851  0.133768   
99950133  -0.166405 -0.273004 -0.270307  -0.271461  ...  -0.188958 -0.338348   
516709288  0.517637  0.285801  0.099218   0.220424  ...   0.247134  0.498477   
36255      0.327220  0.234319  0.099068   0.200317  ...   0.259474  0.239136   
37815204   0.569240  0.198357  0.508204   0.429481  ...   0.537750  0.509776   
...             ...       ...       ...        ...  ...        ...       ...   
27415004   0.589962  0.155899  0.540907   0.484063  ...   0.463150  0.373694   
27415128   0.273169  0.217678  0.586835   0.238365  ...   0.414708  0.336972   
27414937   0.691115  0.348854  0.346977   0.464676  ...   0.390156  0.561629   
66507230  -0.159137 -0.059949 -0.140504  -0.132565  ...   0.003234  0.007918   
100227241 -0.407408 -0.269759 -0.109110  -0.363836  ...  -0.021417 -0.244028   

           15377776  84961467  54000629  27415004  27415128  27414937  \
213688169  0.036309  0.175329 -0.297473  0.121414 -0.088361  0.059033   
99950133   0.059249  0.298978 -0.035843 -0.219705 -0.062054 -0.244888   
516709288 -0.157218 -0.482243  0.246743  0.225187  0.100604  0.526938   
36255     -0.254872 -0.320933  0.081374  0.381771  0.131438  0.341646   
37815204  -0.490047 -0.500738  0.314941  0.842328  0.580039  0.739894   
...             ...       ...       ...       ...       ...       ...   
27415004  -0.562318 -0.395630  0.281569  1.000000  0.485598  0.650653   
27415128  -0.229621 -0.181608  0.226276  0.485598  1.000000  0.474923   
27414937  -0.428574 -0.583834  0.576441  0.650653  0.474923  1.000000   
66507230   0.123305  0.127402 -0.104391 -0.139169 -0.187206  0.029023   
100227241  0.244145  0.352630 -0.479454 -0.266485 -0.285710 -0.593783   

           66507230  100227241  
213688169  0.104548  -0.038691  
99950133  -0.090570  -0.099520  
516709288 -0.021895  -0.227173  
36255     -0.093292  -0.124456  
37815204  -0.059184  -0.320766  
...             ...        ...  
27415004  -0.139169  -0.266485  
27415128  -0.187206  -0.285710  
27414937   0.029023  -0.593783  
66507230   1.000000   0.125457  
100227241  0.125457   1.000000  

[1398 rows x 1398 columns]

In [258]:
features = choix_features(df_ml.corr(),'213688169',0.5)

In [260]:
del features[0]

In [261]:
features

['85147829', '213683992', '30730', '581480367']

In [262]:
X = df_ml.loc[:,features]
Y = df_ml.loc[:,'213688169']

X_train, X_test, y_train, y_test = train_test_split(X, Y, test_size=0.3)

### Régression linéaire avec tuning

In [263]:
# Régression linéaire simple
regressor = LinearRegression()
regressor.fit(X_train, y_train)
y_prediction = regressor.predict(X_test)

print("Le MAE du modèle est "+str(round(mean_absolute_error(y_test, y_prediction),3))+" avec la régression linéaire")

Le MAE du modèle est 1.497 avec la régression linéaire


### Random Forest avec tuning

In [266]:
# Random Forest
regressor = RandomForestRegressor()
regressor.fit(X_train, y_train)
y_prediction = regressor.predict(X_test)

print("Le MAE du modèle est "+str(round(mean_absolute_error(y_test, y_prediction),3))+" avec random forest")

Le MAE du modèle est 0.574 avec random forest


Maintenant, regarder d'autres modèles et essayer sur les différents clusters.
Pour la prochaine fois, [16:05] DEPERSIN Jules
    
Regression où la valeur prédite serait la différence N_velo,(t+1)- N_velo,t plutôt que juste N_vélo,(t+1)
Interpréter les coefficients sur la régression spatiale : LASSO.
Changer l’échelle de temps : 1h
    
Rédiger le modèle proprement sur Latex avec des y_i des x_i. 
